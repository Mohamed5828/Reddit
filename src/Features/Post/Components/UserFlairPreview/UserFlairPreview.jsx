import profile from "Features/Post/Assets/Images/Post-guidelines.png";
import { useState } from "react";
import { HiOutlinePencil } from "react-icons/hi";
import { Separator } from "../CommunityCard/CommunityCard.styled";
import FlairModal from "../FlairModal/FlairModal";
import {
  Container,
  CreatePost,
  DataContainer,
  Flair,
  FlairContainer,
  FlairPreview,
  Name,
  NameAndFlairContainer,
  PreviewContainer,
  ProfilePic,
  ProfilePicContainer,
  UserFlairContainer,
} from "./UserFlairPreview.styled";

// Import api
import axios from "API/axios";
import useFetch from "Hooks/useFetch";

/**
 * component that preview your subreddit flair
 *
 * @returns {React.Component} CreatePost component
 */
const UserFlairPreview = () => {
  //state to show modal
  const [modalShow, setModalShow] = useState(false);

  // State for selected flair index
  const [flairIndex, setFlairIndex] = useState(null);
  /**
   * handler on click on pen handler
   */
  function clickPenHandler() {
    setModalShow(true);
  }

  /**
   * handler when click save on flair modal that takes new flair selected
   *
   * @param {object} flair
   */
  function changeFlair(index) {
    setModalShow(false);
    setFlairIndex(index);
  }

  // Fetch flairs
  const [flairs, error, isLoading, reload] = useFetch({
    axiosInstance: axios,
    method: "GET",
    url: "/flairs/",
    requestConfig: {
      headers: {
        "Content-Language": "en-US",
      },
    },
  });
  console.log(flairs);
  console.log(flairIndex);
  /**
   *
   * @returns {React.Component} CreatePost component
   */
  const Preview = () => {
    return (
      <PreviewContainer>
        <FlairPreview>
          user flair preview
          <button onClick={clickPenHandler}>
            <span className="icon">
              <HiOutlinePencil />
            </span>
          </button>
        </FlairPreview>
        <UserFlairContainer>
          <DataContainer>
            <ProfilePicContainer>
              <ProfilePic profile={profile} />
            </ProfilePicContainer>
            <NameAndFlairContainer>
              <Name>
                <span>user name</span>
              </Name>
              <FlairContainer>
                {flairIndex !== null && (
                  <Flair
                    style={{
                      color: flairs[flairIndex].flairTextColor,
                      backgroundColor: flairs[flairIndex].flairBackGroundColor,
                    }}
                  >
                    {flairs[flairIndex].text}
                  </Flair>
                )}
              </FlairContainer>
            </NameAndFlairContainer>
          </DataContainer>
        </UserFlairContainer>
      </PreviewContainer>
    );
  };

  return (
    <Container>
      {/* dynamic display cuz if there is no display */}
      <Preview />
      <FlairModal
        show={modalShow}
        onHide={() => setModalShow(false)}
        flairList={flairs}
        flairIndex={flairIndex}
        setFlairIndex={setFlairIndex}
        error={error}
        isLoading={isLoading}
        postOrUser={"user"}
      />
    </Container>
  );
};

export default UserFlairPreview;
