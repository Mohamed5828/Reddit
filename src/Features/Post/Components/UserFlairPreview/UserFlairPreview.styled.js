import { Link } from "react-router-dom";
// ${Link} {}
import styled from "styled-components";

export const Container = styled.div`
  margin-top: 12px;
`;

export const CreatePost = styled(Link)`
  display: flex;
  /* text-align: center; */
  text-decoration: none;
  width: 100%;
  box-sizing: border-box;
  border-radius: 20px;
  align-items: center;
  justify-content: center;
  font-family: "Noto Sans", Arial, sans-serif;
  font-size: 14px;
  font-weight: 700;
  padding: 4px 16px;
  position: relative;
  background-color: #0079d3;
  border: none;
  color: #fff;

  :hover {
    background-color: #1483d6;
    color: #fff;
  }
`;

export const PreviewContainer = styled.div`
  margin-top: 12px;
`;

export const FlairPreview = styled.div`
  display: flex;
  font-size: 10px;
  font-weight: 700;
  letter-spacing: 0.5px;
  line-height: 12px;
  text-transform: uppercase;
  justify-content: space-between;
  align-items: center;

  button {
    padding: 0;
    display: flex;
    justify-content: center;
    text-align: center;
    position: relative;
    border: 1px solid transparent;
    color: #0079d3;
    background: transparent;
    border-radius: 50%;
    min-height: 25px;
    min-width: 25px;
    width: auto;
    font-family: "Noto Sans", Arial, sans-serif;
    font-size: 14px;
    font-weight: 700;
    line-height: 17px;
    margin-left: auto;
  }

  button:hover {
    background-color: #ececec;
  }

  span.icon {
    font-size: 20px;
    font-weight: 400;
    height: 20px;
    line-height: 20px;
    width: 20px;
  }
`;

export const UserFlairContainer = styled.div`
  display: flex;
  margin: 8px 0;
  width: 100%;
`;

export const DataContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  max-width: 100;
`;

export const ProfilePicContainer = styled.div`
  height: 100%;
  position: relative;
`;
export const ProfilePic = styled.div`
  height: 30px;
  width: 30px;
  margin-top: 6px;
  background: #f9c112 url(${(props) => props.profile}) no-repeat center top;
  background-size: cover;
  border-radius: 50%;
`;

export const NameAndFlairContainer = styled.div`
  display: flex;
  align-items: flex-start;
  flex-direction: column;
  margin-left: 8px;
  min-width: 0;
`;

export const Name = styled.div`
  display: flex;
  align-items: center;
  width: 100%;

  span {
    font-size: 12px;
    font-weight: 500;
    line-height: 16px;
    color: ${({ theme }) => theme.color.primary};
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }
`;

export const FlairContainer = styled.div`
  display: flex;
  align-items: center;
  width: 100%;
  max-width: 100%;
  margin-top: 2px;
`;

export const Flair = styled.span`
  font-weight: 400;
  box-sizing: border-box;
  font-size: 12px;
  line-height: 16px;
  border-radius: 2px;
  display: inline-block;
  margin-right: 5px;
  overflow: hidden;
  text-overflow: ellipsis;
  vertical-align: text-bottom;
  white-space: pre;
  word-break: normal;
  padding: 0 4px;
`;
