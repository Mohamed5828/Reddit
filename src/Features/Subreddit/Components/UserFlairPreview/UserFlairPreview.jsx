import profile from "Features/Subreddit/Assets/index.jpeg";
import { useState } from "react";
import { HiOutlinePencil } from "react-icons/hi";
import { Separator } from "../CommunityCard/CommunityCard.styled";
import SelectYourFlair from "../SelectYourFlair/SelectYourFlair";
import {
  Container,
  CreatePost,
  DataContainer,
  Flair,
  FlairContainer,
  FlairPreview,
  Name,
  NameAndFlairContainer,
  PreviewContainer,
  ProfilePic,
  ProfilePicContainer,
  UserFlairContainer
} from "./UserFlairPreview.styled";

/**
 * component that preview your subreddit flair
 * 
 * @returns {React.Component}
 */
const UserFlairPreview = () => {
  // const flair = { text: "", backgroundColor: "", color: "" };
  const [flair, setFlair] = useState({
    text: "hi",
    color: "white",
    backgroundColor: "#000",
  });
  //state to show modal
  const [modalShow, setModalShow] = useState(false);

  /**
   * handler on click on pen handler
   */
  function clickPenHandler() {
    setModalShow(true);
  }
  
  /**
   * handler when click save on flair modal that takes new flair selected 
   * 
   * @param {object} flair 
   */
  function changeFlair(flair) {
    setModalShow(false);
    setFlair(flair);
  }

  /**
   * component preview your flair in this subreddit
   * 
   * @returns {React.Component} 
   */
  const Preview = () => {
    return (
      <PreviewContainer>
        <FlairPreview>
          user flair preview
          <button onClick={clickPenHandler}>
            <span className="icon">
              <HiOutlinePencil />
            </span>
          </button>
        </FlairPreview>
        <UserFlairContainer>
          <DataContainer>
            <ProfilePicContainer>
              <ProfilePic profile={profile} />
            </ProfilePicContainer>
            <NameAndFlairContainer>
              <Name>
                <span>user name</span>
              </Name>
              <FlairContainer>
                <Flair
                  style={{
                    color: flair.color,
                    backgroundColor: flair.backgroundColor,
                  }}
                >
                  {flair.text}
                </Flair>
              </FlairContainer>
            </NameAndFlairContainer>
          </DataContainer>
        </UserFlairContainer>
      </PreviewContainer>
    );
  };

  return (
    <Container>
      <CreatePost to="/submit/" className="create-post">Create Post</CreatePost>
      {/* dynamic display cuz if there is no display */}
      {true && <Separator />}
      {true && <Preview />}
      <SelectYourFlair
        modalShow={modalShow}
        hide={() => setModalShow(false)}
        onApply={changeFlair}
      />
    </Container>
  );
};

export default UserFlairPreview;
