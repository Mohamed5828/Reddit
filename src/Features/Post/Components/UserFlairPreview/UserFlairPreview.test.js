import Enzyme, { shallow } from "enzyme";
import UserFlairPreview from "./UserFlairPreview";

describe("preview of selected flair", () => {
  it("should render without crashing", () => {
    expect(shallow(<UserFlairPreview />)).toMatchSnapshot();
  });
});
