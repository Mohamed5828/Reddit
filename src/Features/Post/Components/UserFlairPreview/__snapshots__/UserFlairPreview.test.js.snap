// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preview of selected flair should render without crashing 1`] = `
<styled.div>
  <Preview />
  <FlairModal
    error=""
    flairIndex={null}
    flairList={Array []}
    isLoading={true}
    onHide={[Function]}
    postOrUser="user"
    setFlairIndex={[Function]}
    show={false}
  />
</styled.div>
`;
